### PR TITLE
return architecture levels for micromamba

### PR DESCRIPTION
**Pull request by isuruf**
cc @beckermr, @h-vetinari, @jaimergp, @wolfv

### CODE REVIEW

File: libmamba/src/core/virtual_packages.cpp [hunk 1]
```diff
@@ -163,6 +163,30 @@ namespace mamba
...
+        {
+            return "x86_64_v4";
+        }
+#else
+        std::string get_archspec_x86_64()
+        {
+            return "x86_64";
+        }
+#endif
+
```
Comment (mbargull) [inline libmamba/src/core/virtual_packages.cpp:189]: There can be corner cases like Intel's Penryn processors that support SSE4.1 but not POPCNT. Hence, it is probably better to check for more features.
Acc. to
https://en.wikipedia.org/wiki/X86-64#Microarchitecture_levels
and
https://github.com/bminor/glibc/blob/glibc-2.38/sysdeps/x86/get-isa-level.h
the following is not exhaustive but close enough:
```suggestion
        std::string get_archspec_x86_64()
        {
#if (defined(__GNUC__) || defined(__clang__)) && __x86_64__
            /* if (__builtin_cpu_supports ("x86-64-v4")) */
            if (__builtin_cpu_supports("avx512f") && __builtin_cpu_supports("avx512bw")
                && __builtin_cpu_supports("avx512cd") && __builtin_cpu_supports("avx512dq")
                && __builtin_cpu_supports("avx512vl"))
            {
                return "x86_64-v4";
            }
            /* if (__builtin_cpu_supports ("x86-64-v3")) */
            if (__builtin_cpu_supports("avx") && __builtin_cpu_supports("avx2")
                && __builtin_cpu_supports("bmi") && __builtin_cpu_supports("bmi2")
                && __builtin_cpu_supports("fma"))
            {
                return "x86_64-v3";
            }
            /* if (__builtin_cpu_supports ("x86-64-v2")) */
            if (__builtin_cpu_supports("popcnt") && __builtin_cpu_supports("sse3")
                && __builtin_cpu_supports("ssse3") && __builtin_cpu_supports("sse4.1")
                && __builtin_cpu_supports("sse4.2"))
            {
                return "x86_64-v2";
            }
#endif
            return "x86_64";
        }
```
With GCC 12+ and future Clang versions, `__builtin_cpu_supports ("x86-64-v2")` et al., can be called directly.

(Haven't tested this on macOS.)

Comment (isuruf) [reply libmamba/src/core/virtual_packages.cpp]: With the function multiversioning code I had, we could also check for multiple CPU features by repeating the target. However using `__builtin_cpu_supports` seems less magic-y than function multi-versioning.

Comment (h-vetinari) [reply libmamba/src/core/virtual_packages.cpp]: > and future Clang versions

Surprised this landed so late, but yeah: https://github.com/llvm/llvm-project/commit/27da15381cbe2ac6fd1319f6409dbbab9a857b7b

Comment (mbargull) [reply libmamba/src/core/virtual_packages.cpp]: > With the function multiversioning code I had, we could also check for multiple CPU features by repeating the target. However using `__builtin_cpu_supports` seems less magic-y than function multi-versioning.

Oh, I thought repeating targets would be disjunctive not conjunctive. Maybe I'm just missing something.
In other context I like the multiversioning better, too ;). GCC 12 (Clang not yet) also seems to accept `__attribute__((target ("arch=x86-64-v3")))` which is also nice.

Comment (isuruf) [reply libmamba/src/core/virtual_packages.cpp]: GCC and clang internally have a total-ordering of the microarchitecture (unlike archspec which has a partial ordering only) and selects a minimum micro-arch based on the targets given and checks at runtime if the runtime micro-arch is equal or higher.

Comment (mbargull) [reply libmamba/src/core/virtual_packages.cpp]: Ah, alright, thanks! :bow:

**Comment by jonashaag** [comment]
Is there a way to unit test this?

**Comment by JohanMabille** [comment]
> Is there a way to unit test this?

We could check preprocessor tokens (like we do in [xsimd](https://github.com/xtensor-stack/xsimd/blob/master/include/xsimd/config/xsimd_config.hpp)) to detect the architecture, and check that result of  `get_archspec_x86_64` is consistent, but that would test only the architecture available on the CI. If we want to be exhaustive, we could explictly pass the arch we want ot use in the CI (see [xsimd again ;)](https://github.com/xtensor-stack/xsimd/blob/master/.github/workflows/linux.yml#L16) for an example), but that would require a lot of new jobs in the CI, I'm not sure we want that for testing a single function given the tiime the CI is already taking.

**Review by JohanMabille — APPROVED** [review]

**Comment by jonashaag** [comment]
Maybe we can at least compare the output of the Mamba logic to the output of the xsimd logic? That's still limited to the runner architecture but it's better than nothing

**Comment by mbargull** [comment]
On Linux you could use QEMU. E.g., if you were to compile a test program that prints this function's output:
```sh
[ "$( qemu-x86_64 -cpu Penryn test-get_archspec_x86_64` )" = "x86_64" ]
[ "$( qemu-x86_64 -cpu Nehalem test-get_archspec_x86_64` )" = "x86_64-v2" ]
[ "$( qemu-x86_64 -cpu Haswell test-get_archspec_x86_64` )" = "x86_64-v3" ]
[ "$( qemu-x86_64 -cpu Skylake-Server test-get_archspec_x86_64` )" = "x86_64-v4" ]
```
This would only work if the host CPU supports the features too. I.e., depending on the runner, you may want to guard the test for `Skylake-Server` at least.
And this would only test on Linux. No idea what the emulators on macOS offers.

**Comment by isuruf** [comment]
I think adding a better test would not be worth the trouble. If we use `__builtin_cpu_supports` in the test, we'll be duplicating the actual code in test code which doesn't really test anything. If we use preprocessor macros like __AVX2__, we need to set -march=native, but we need quite a bit of logic to make that work. (Check that it is UNIX and (clang or gcc) and x86_64 which is not easy in CMake as there are options like CMAKE_OSX_ARCHITECTURES)
Finally using an emulator like QEMU or Intel SDE would work, but not sure it's worth it to complicate the CI infrastructure.

**Pull request merged**